### PR TITLE
Use run.Group to run goroutines in controller's main.go

### PR DESCRIPTION
**Pull request by metalmatze**
> run.Group is a universal mechanism to manage goroutine lifecycles.
Please take a look at https://github.com/oklog/run

A quick example:
```
ln, _ := net.Listen("tcp", ":8080")
g.Add(func() error {
	return http.Serve(ln, nil)
}, func(error) {
	ln.Close()
})
```
> Actors are defined as a pair of functions: an **execute** function, which should run synchronously; and an **interrupt** function, which, when invoked, should cause the execute function to return. Finally, invoke Run, which blocks until the first actor returns. This general-purpose API allows callers to model pretty much any runnable task, and achieve well-defined lifecycle semantics for the group.

So what happens is, that we now run 3 actors. One that listens to signals, one that runs the internal http server and the last one that runs the controller. If either of the 3 returns the others are shutdown.

This makes it way easier to reason about the structure of concurrently running things in the main.go

### CODE REVIEW

File: api/cmd/kubermatic-cluster-controller/main.go [hunk 1]
```diff
@@ -73,75 +78,118 @@ func main() {
...
+				go cache.WaitForCacheSync(stop, seedInformerGroup.HasSynced)
+
+				glog.Info("Starting controller")
+				go ctrl.Run(*workerCount, stop)
+			}
+			<-stop
+			return nil
+		}, func(err error) {
+			glog.Info("Stopping controllers")
+			stop <- struct{}{}
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-cluster-controller/main.go]: Close the `stop` channel instead

Comment (metalmatze) [reply api/cmd/kubermatic-cluster-controller/main.go]: Ah! Good catch! :grin:

**Comment by metalmatze** [comment]
```
^C
I1124 12:13:08.931271    4833 main.go:115] Shutting down the internal http server
I1124 12:13:08.931370    4833 main.go:189] Stopping controllers
```
:tada:

File: api/cmd/kubermatic-cluster-controller/main.go [hunk 2]
```diff
@@ -73,75 +78,118 @@ func main() {
...
+				glog.Error("failed to shutdown the internal http server gracefully:", err)
+			}
+		})
+	}
+	// This group is running the actual controller logic
+	{
+		stop := make(chan struct{})
+
+		g.Add(func() error {
+			for dc := range clientcmdConfig.Contexts {
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-cluster-controller/main.go]: Can we rather add each controller instance to the group?

We then could use the blocking `ctrl.Run()` which then really tells us that the controller finished when it returns

Comment (metalmatze) [reply api/cmd/kubermatic-cluster-controller/main.go]: I tried that yesterday, I think. I can try again.

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**